File: apps/desktop/src/main/defaults.test.ts
import assert from "node:assert/strict";
import test from "node:test";
import {
  initializeDesktopEnvironment,
  resolveDesktopDefaultsFromEnv,
  resolveDesktopUserDataPath,
  resolveTuttiEnv
} from "./defaults.ts";

test("resolveDesktopDefaultsFromEnv uses generated development defaults", () => {
  const previousEnv = { ...process.env };
  const homeDir = "/tmp/tutti-desktop-home";

  try {
    process.env.HOME = homeDir;
    process.env.TUTTI_ENV = "development";
    delete process.env.TUTTI_STATE_DIR;
    delete process.env.TUTTI_LOG_DIR;
    delete process.env.TUTTID_RUN_DIR;
    delete process.env.TUTTID_LOG_PATH;
    delete process.env.TUTTI_DESKTOP_LOG_PATH;
    delete process.env.TUTTID_LISTENER_INFO_PATH;
    delete process.env.TUTTID_ADDR;

    const got = resolveDesktopDefaultsFromEnv();

    assert.equal(got.runtime.env, "development");
    assert.equal(got.state.rootDir, `${homeDir}/.tutti-dev`);
    assert.equal(got.state.logsDir, `${homeDir}/.tutti-dev/logs`);
    assert.equal(got.state.runDir, `${homeDir}/.tutti-dev/run`);
    assert.equal(got.state.tuttidDBPath, `${homeDir}/.tutti-dev/tuttid.db`);
    assert.equal(
      got.state.tuttidListenerInfoPath,
      `${homeDir}/.tutti-dev/run/tuttid.listener.json`
    );
    assert.equal(
      got.state.tuttidLogPath,
      `${homeDir}/.tutti-dev/logs/tuttid.log`
    );
    assert.equal(
      got.state.desktopLogPath,
      `${homeDir}/.tutti-dev/logs/tutti-desktop.log`
    );
    assert.equal(
      got.state.tuttidPIDPath,
      `${homeDir}/.tutti-dev/run/tuttid.pid`
    );
    assert.equal(got.transport.tcpAddr, "127.0.0.1:4545");
    assert.equal(got.logging.defaultLevel, "info");
    assert.equal(got.logging.defaultOutput, "file");
    assert.equal(got.logging.maxSizeMB, 50);
    assert.equal(got.logging.maxBackups, 10);
    assert.equal(got.logging.maxAgeDays, 14);
    assert.equal(got.logging.maxTotalMB, 300);
  } finally {
    restoreEnv(previousEnv);
  }
});

test("resolveDesktopDefaultsFromEnv honors endpoint and log overrides", () => {
  const previousEnv = { ...process.env };

  try {
    process.env.TUTTI_ENV = "production";
    process.env.TUTTI_LOG_DIR = "/tmp/tutti-logs";
    process.env.TUTTID_ADDR = "127.0.0.1:9999";
    process.env.TUTTID_LISTENER_INFO_PATH = "/tmp/tuttid.listener.json";

    const got = resolveDesktopDefaultsFromEnv();

    assert.equal(got.transport.tcpAddr, "127.0.0.1:9999");
    assert.equal(got.state.tuttidListenerInfoPath, "/tmp/tuttid.listener.json");
    assert.equal(got.state.logsDir, "/tmp/tutti-logs");
    assert.equal(got.state.tuttidLogPath, "/tmp/tutti-logs/tuttid.log");
    assert.equal(got.state.desktopLogPath, "/tmp/tutti-logs/tutti-desktop.log");
  } finally {
    restoreEnv(previousEnv);
  }
});

test("initializeDesktopEnvironment sets development and production defaults when unset", () => {
  const previousEnv = { ...process.env };

  try {
    delete process.env.TUTTI_ENV;
    initializeDesktopEnvironment({ isPackaged: false });
    assert.equal(resolveTuttiEnv(), "development");

    delete process.env.TUTTI_ENV;
    initializeDesktopEnvironment({ isPackaged: true });
    assert.equal(resolveTuttiEnv(), "production");
  } finally {
    restoreEnv(previousEnv);
  }
});

test("initializeDesktopEnvironment sets shared app version when unset", () => {
  const previousEnv = { ...process.env };

  try {
    delete process.env.TUTTI_APP_VERSION;

    initializeDesktopEnvironment({
      appVersion: "1.2.3",
      isPackaged: false
    });

    assert.equal(process.env.TUTTI_APP_VERSION, "1.2.3");
  } finally {
    restoreEnv(previousEnv);
  }
});

test("initializeDesktopEnvironment preserves shared app version override", () => {
  const previousEnv = { ...process.env };

  try {
    process.env.TUTTI_APP_VERSION = "9.9.9";

    initializeDesktopEnvironment({
      appVersion: "1.2.3",
      isPackaged: false
    });

    assert.equal(process.env.TUTTI_APP_VERSION, "9.9.9");
  } finally {
    restoreEnv(previousEnv);
  }
});

test("resolveDesktopUserDataPath isolates development Electron storage", () => {
  const previousEnv = { ...process.env };

  try {
    process.env.TUTTI_ENV = "development";

    assert.equal(
      resolveDesktopUserDataPath({
        appDataDir: "/tmp/app-data",
        appName: "Tutti"
      }),
      "/tmp/app-data/Tutti-dev"
    );
  } finally {
    restoreEnv(previousEnv);
  }
});

test("resolveDesktopUserDataPath keeps production on Electron defaults", () => {
  const previousEnv = { ...process.env };

  try {
    process.env.TUTTI_ENV = "production";

    assert.equal(
      resolveDesktopUserDataPath({
        appDataDir: "/tmp/app-data",
        appName: "Tutti"
      }),
      null
    );
  } finally {
    restoreEnv(previousEnv);
  }
});

function restoreEnv(previousEnv: NodeJS.ProcessEnv): void {
  for (const key of Object.keys(process.env)) {
    if (!(key in previousEnv)) {
      delete process.env[key];
    }
  }

  for (const [key, value] of Object.entries(previousEnv)) {
    if (value === undefined) {
      delete process.env[key];
      continue;
    }

    process.env[key] = value;
  }
}

File: services/tuttid/types/defaults_test.go
package types

import (
	"path/filepath"
	"testing"
)

func TestResolveDefaultsFromEnvUsesSharedGeneratedDefaults(t *testing.T) {
	homeDir := t.TempDir()
	t.Setenv("HOME", homeDir)
	t.Setenv("TUTTI_ENV", "development")
	t.Setenv("TUTTI_STATE_DIR", "")
	t.Setenv("TUTTI_LOG_DIR", "")
	t.Setenv("TUTTID_ADDR", "")
	t.Setenv("TUTTID_LISTENER_INFO_PATH", "")

	got := ResolveDefaultsFromEnv()

	assertEqual(t, got.Runtime.Env, "development")
	assertEqual(t, got.State.RootDir, filepath.Join(homeDir, ".tutti-dev"))
	assertEqual(t, got.State.LogsDir, filepath.Join(homeDir, ".tutti-dev", "logs"))
	assertEqual(t, got.State.RunDir, filepath.Join(homeDir, ".tutti-dev", "run"))
	assertEqual(t, got.State.TuttidDBPath, filepath.Join(homeDir, ".tutti-dev", "tuttid.db"))
	assertEqual(t, got.State.TuttidListenerInfoPath, filepath.Join(homeDir, ".tutti-dev", "run", "tuttid.listener.json"))
	assertEqual(t, got.State.TuttidLogPath, filepath.Join(homeDir, ".tutti-dev", "logs", "tuttid.log"))
	assertEqual(t, got.State.DesktopLogPath, filepath.Join(homeDir, ".tutti-dev", "logs", "tutti-desktop.log"))
	assertEqual(t, got.State.TuttidPIDPath, filepath.Join(homeDir, ".tutti-dev", "run", "tuttid.pid"))
	assertEqual(t, got.Transport.TCPAddr, "127.0.0.1:4545")
	assertEqual(t, got.Logging.DefaultLevel, "info")
	assertEqual(t, got.Logging.DefaultOutput, "file")
	assertEqualInt(t, got.Logging.MaxSizeMB, 50)
	assertEqualInt(t, got.Logging.MaxBackups, 10)
	assertEqualInt(t, got.Logging.MaxAgeDays, 14)
	assertEqualInt(t, got.Logging.MaxTotalMB, 300)
}

func TestResolveDefaultsFromEnvAppliesOverrides(t *testing.T) {
	t.Setenv("TUTTI_ENV", "development")
	t.Setenv("TUTTI_STATE_DIR", "/tmp/tutti-state")
	t.Setenv("TUTTI_LOG_DIR", "/tmp/tutti-logs")
	t.Setenv("TUTTID_ADDR", "127.0.0.1:1111")

	got := ResolveDefaultsFromEnv()

	assertEqual(t, got.Runtime.Env, "development")
	assertEqual(t, got.State.RootDir, "/tmp/tutti-state")
	assertEqual(t, got.State.LogsDir, "/tmp/tutti-logs")
	assertEqual(t, got.Transport.TCPAddr, "127.0.0.1:1111")
}

func TestDesktopLoginCallbackURLUsesEnvironmentScheme(t *testing.T) {
	t.Setenv("TUTTI_ENV", "development")
	assertEqual(t, DesktopLoginCallbackURL(), "tutti-dev://login/callback")

	t.Setenv("TUTTI_ENV", "production")
	assertEqual(t, DesktopLoginCallbackURL(), "tutti://login/callback")
}

func TestResolveAnalyticsConfigUsesGeneratedDefaults(t *testing.T) {
	t.Setenv("TUTTI_ENV", "")
	t.Setenv("TUTTI_ANALYTICS_DISABLED", "")
	t.Setenv("TUTTI_ANALYTICS_APP_ID", "")
	t.Setenv("TUTTI_ANALYTICS_APP_KEY", "")
	t.Setenv("TUTTI_ANALYTICS_CHANNEL_DOMAIN", "")
	t.Setenv("TUTTI_APP_VERSION", "")
	t.Setenv("TUTTI_ANALYTICS_APP_VERSION", "")

	got := ResolveAnalyticsConfig()

	if got.Disabled {
		t.Fatal("analytics disabled = true, want false")
	}
	assertEqualInt(t, got.AppID, 20004134)
	assertEqual(t, got.AppKey, "984646081c1dc9dbe502e9c5e17711fbf9d9fdb85047eb7808db4776c34c0af0")
	assertEqual(t, got.Channel, "sg")
	assertEqual(t, got.ChannelDomain, "https://gator.uba.ap-southeast-1.volces.com")
	assertEqual(t, got.AppVersion, "0.0.0")
}

func TestResolveAnalyticsConfigEnablesDebugPipelineInDevelopment(t *testing.T) {
	t.Setenv("TUTTI_ENV", "development")
	t.Setenv("TUTTI_ANALYTICS_DISABLED", "")
	t.Setenv("TUTTI_ANALYTICS_APP_ID", "")
	t.Setenv("TUTTI_ANALYTICS_APP_KEY", "")
	t.Setenv("TUTTI_ANALYTICS_CHANNEL_DOMAIN", "")
	t.Setenv("TUTTI_APP_VERSION", "")
	t.Setenv("TUTTI_ANALYTICS_APP_VERSION", "")

	got := ResolveAnalyticsConfig()

	if got.Disabled {
		t.Fatal("analytics disabled = true, want false so development can publish local debug events")
	}
	if !got.Debug {
		t.Fatal("analytics debug = false, want true in development")
	}
}

func TestResolveAnalyticsConfigAppliesOverrides(t *testing.T) {
	t.Setenv("TUTTI_ENV", "")
	t.Setenv("TUTTI_ANALYTICS_DISABLED", "true")
	t.Setenv("TUTTI_ANALYTICS_APP_ID", "123")
	t.Setenv("TUTTI_ANALYTICS_APP_KEY", "dev-key")
	t.Setenv("TUTTI_ANALYTICS_CHANNEL_DOMAIN", "https://example.test")
	t.Setenv("TUTTI_APP_VERSION", "")
	t.Setenv("TUTTI_ANALYTICS_APP_VERSION", "1.2.3")

	got := ResolveAnalyticsConfig()

	if !got.Disabled {
		t.Fatal("analytics disabled = false, want true")
	}
	assertEqualInt(t, got.AppID, 123)
	assertEqual(t, got.AppKey, "dev-key")
	assertEqual(t, got.ChannelDomain, "https://example.test")
	assertEqual(t, got.AppVersion, "1.2.3")
}

func TestResolveAnalyticsConfigEnablesDebugForDevelopmentOnly(t *testing.T) {
	cases := []struct {
		name     string
		tuttiEnv string
		want     bool
	}{
		{name: "development", tuttiEnv: "development", want: true},
		{name: "production", tuttiEnv: "production", want: false},
		{name: "unset env", tuttiEnv: "", want: false},
	}

	for _, tt := range cases {
		t.Run(tt.name, func(t *testing.T) {
			t.Setenv("TUTTI_ENV", tt.tuttiEnv)
			t.Setenv("TUTTI_ANALYTICS_DISABLED", "")
			t.Setenv("TUTTI_ANALYTICS_APP_ID", "")
			t.Setenv("TUTTI_ANALYTICS_APP_KEY", "")
			t.Setenv("TUTTI_ANALYTICS_CHANNEL_DOMAIN", "")
			t.Setenv("TUTTI_APP_VERSION", "")
			t.Setenv("TUTTI_ANALYTICS_APP_VERSION", "")

			got := ResolveAnalyticsConfig()

			assertEqualBool(t, got.Debug, tt.want)
		})
	}
}

func TestResolveAnalyticsConfigParsesDisabledOverride(t *testing.T) {
	tests := []struct {
		name  string
		value string
		want  bool
	}{
		{name: "empty", value: "", want: false},
		{name: "true", value: "true", want: true},
		{name: "uppercase true", value: "TRUE", want: true},
		{name: "one", value: "1", want: true},
		{name: "yes", value: "yes", want: true},
		{name: "false", value: "false", want: false},
		{name: "uppercase false", value: "FALSE", want: false},
		{name: "zero", value: "0", want: false},
		{name: "no", value: "no", want: false},
		{name: "invalid", value: "treu", want: true},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			t.Setenv("TUTTI_ANALYTICS_DISABLED", tt.value)
			t.Setenv("TUTTI_ANALYTICS_APP_ID", "")
			t.Setenv("TUTTI_ANALYTICS_APP_KEY", "")
			t.Setenv("TUTTI_ANALYTICS_CHANNEL_DOMAIN", "")
			t.Setenv("TUTTI_APP_VERSION", "")
			t.Setenv("TUTTI_ANALYTICS_APP_VERSION", "")

			got := ResolveAnalyticsConfig()

			assertEqualBool(t, got.Disabled, tt.want)
		})
	}
}

func TestResolveAnalyticsConfigParsesAppIDOverride(t *testing.T) {
	tests := []struct {
		name  string
		value string
		want  int
	}{
		{name: "empty", value: "", want: 20004134},
		{name: "positive integer", value: "123", want: 123},
		{name: "zero", value: "0", want: 0},
		{name: "negative", value: "-1", want: 0},
		{name: "non numeric", value: "abc", want: 0},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			t.Setenv("TUTTI_ANALYTICS_DISABLED", "")
			t.Setenv("TUTTI_ANALYTICS_APP_ID", tt.value)
			t.Setenv("TUTTI_ANALYTICS_APP_KEY", "")
			t.Setenv("TUTTI_ANALYTICS_CHANNEL_DOMAIN", "")
			t.Setenv("TUTTI_APP_VERSION", "")
			t.Setenv("TUTTI_ANALYTICS_APP_VERSION", "")

			got := ResolveAnalyticsConfig()

			assertEqualInt(t, got.AppID, tt.want)
		})
	}
}

func TestResolveAnalyticsConfigResolvesAppVersionOverride(t *testing.T) {
	tests := []struct {
		name                     string
		appVersion               string
		analyticsSpecificVersion string
		want                     string
	}{
		{name: "empty", appVersion: "", analyticsSpecificVersion: "", want: "0.0.0"},
		{name: "shared app version", appVersion: "1.2.3", analyticsSpecificVersion: "", want: "1.2.3"},
		{name: "analytics specific version takes precedence", appVersion: "1.2.3", analyticsSpecificVersion: "2.3.4", want: "2.3.4"},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			t.Setenv("TUTTI_ANALYTICS_DISABLED", "")
			t.Setenv("TUTTI_ANALYTICS_APP_ID", "")
			t.Setenv("TUTTI_ANALYTICS_APP_KEY", "")
			t.Setenv("TUTTI_ANALYTICS_CHANNEL_DOMAIN", "")
			t.Setenv("TUTTI_APP_VERSION", tt.appVersion)
			t.Setenv("TUTTI_ANALYTICS_APP_VERSION", tt.analyticsSpecificVersion)

			got := ResolveAnalyticsConfig()

			assertEqual(t, got.AppVersion, tt.want)
		})
	}
}

func assertEqual(t *testing.T, got string, want string) {
	t.Helper()
	if got != want {
		t.Fatalf("got %q, want %q", got, want)
	}
}

func assertEqualInt(t *testing.T, got int, want int) {
	t.Helper()
	if got != want {
		t.Fatalf("got %d, want %d", got, want)
	}
}

func assertEqualBool(t *testing.T, got bool, want bool) {
	t.Helper()
	if got != want {
		t.Fatalf("got %t, want %t", got, want)
	}
}
